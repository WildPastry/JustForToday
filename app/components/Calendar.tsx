import { ICalendar, IDayItem, IMonthItem } from '../types/date.types';
import { Pressable, StyleSheet } from 'react-native';
import { Text, View } from '../components/Themed';
import { useEffect, useState } from 'react';
import { AppState } from '../redux/store';
import DayItem from '../components/DayItem';
import { MonoText } from '../components/StyledText';
import MonthItem from '../components/MonthItem';
import { useAppSelector } from '../redux/hooks';

const Calendar: React.FC<ICalendar> = ({
  handleCalendarChange
}: ICalendar): JSX.Element => {
  // App selector for month data
  const monthItems = useAppSelector((state: AppState): IMonthItem[] => {
    return state.data.monthItems;
  });

  // Local data
  const [months, setMonths] = useState<IMonthItem[]>([]);
  const [days, setDays] = useState<IDayItem[]>([]);

  // Effect for setting the month items
  useEffect(() => {
    setMonths(monthItems);
  }, [monthItems]);

  // Reset to all months
  const getAllMonths = (): void => {
    setMonths(monthItems);
    setDays([]);
  };

  // Handling click functions
  const handleMonthClick = (month: IMonthItem): void => {
    setMonths([month]);
    setDays(month.days);
  };

  const handleDayClick = (day: IDayItem): void => {
    setDays([day]);
    handleCalendarChange(false, day.id);
  };

  return (
    <View style={styles.container}>
      <MonoText style={styles.title}>Calendar</MonoText>
      {/* Reset to all months */}
      <Pressable onPress={() => getAllMonths()}>
        <Text>ALL MONTHS</Text>
      </Pressable>
      {/* Months */}
      {months.map((month, index) => (
        <MonthItem
          key={index}
          id={month.id}
          name={month.name}
          days={month.days}
          onPress={() => handleMonthClick(month)}
        />
      ))}
      {/* Days */}
      {days.map((day, index) => (
        <DayItem
          key={index}
          id={day.id}
          name={day.name}
          onPress={() => handleDayClick(day)}
        />
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
    padding: 20
  },
  title: {
    fontSize: 20
  }
});

export default Calendar;
