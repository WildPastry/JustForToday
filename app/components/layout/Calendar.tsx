import { ICalendar, IDayItem, IMonthItem } from '../../types/date.types';
import {
  FlatList,
  Pressable,
  StyleSheet,
  TouchableOpacity
} from 'react-native';
import { ScrollView, Text, View } from '../styles/Themed';
import { setSelectedDay, setSelectedMonth } from '../../redux/slices/dateSlice';
import { useAppDispatch, useAppSelector } from '../../redux/hooks';
import { useEffect, useState } from 'react';

import { AppState } from '../../redux/store';
import DayItem from './DayItem';
import { FontDisplay } from '../styles/StyledText';
import { MaterialIcons } from '@expo/vector-icons';
import MonthItem from './MonthItem';

// FLATLIST STUFF
type ItemData = {
  id: string;
  title: string;
};

const DATA: ItemData[] = [
  {
    id: 'bd7acbea-c1b1-46c2-aed5-3ad53abb28ba',
    title: '1'
  },
  {
    id: '3ac68afc-c605-48d3-a4f8-fbd91aa97f63',
    title: '2'
  },
  {
    id: '58694a0f-3da1-471f-bd96-145571e29d72',
    title: '3'
  }
];

type ItemProps = {
  item: ItemData;
  onPress: () => void;
  backgroundColor: string;
  textColor: string;
};

const Item = ({ item, onPress, backgroundColor, textColor }: ItemProps) => (
  <TouchableOpacity
    onPress={onPress}
    style={[styles.item, { backgroundColor }]}>
    <Text style={[styles.itemText, { color: textColor }]}>{item.title}</Text>
  </TouchableOpacity>
);

const Calendar: React.FC<ICalendar> = (props: ICalendar): JSX.Element => {
  // FLATLIST STUFF
  const [selectedId, setSelectedId] = useState<string>();

  const renderItem = ({ item }: { item: ItemData }) => {
    const backgroundColor = item.id === selectedId ? '#6e3b6e' : '#f9c2ff';
    const color = item.id === selectedId ? 'white' : 'black';

    return (
      <Item
        item={item}
        onPress={() => setSelectedId(item.id)}
        backgroundColor={backgroundColor}
        textColor={color}
      />
    );
  };

  // Component settings
  const dispatch = useAppDispatch();
  const [months, setMonths] = useState<IMonthItem[]>([]);
  const [days, setDays] = useState<IDayItem[]>([]);

  // Data from store
  const monthItems: IMonthItem[] = useAppSelector(
    (state: AppState): IMonthItem[] => {
      return state.data.monthItems;
    }
  );

  useEffect(() => {
    setMonths(monthItems);
  }, [monthItems]);

  const getAllMonths = (): void => {
    setMonths(monthItems);
    setDays([]);
  };

  const handleMonth = (month: IMonthItem): void => {
    dispatch(setSelectedMonth(month.id));
    setMonths([month]);
    setDays(month.days);
  };

  const handleDay = (day: IDayItem): void => {
    dispatch(setSelectedDay(day.id));
    setDays([day]);
    props.handleCalendarChange(false, day.id);
  };

  const toggleCalendar = (): void => {
    props.handleCalendarState(false);
  };

  return (
    <View style={styles.container}>

<FlatList
        numColumns={3}
        data={DATA}
        renderItem={renderItem}
        keyExtractor={(item) => item.id}
        extraData={selectedId}
      />

      {/* Controls */}
      <View style={styles.controls}>
        <Pressable style={styles.backIcon} onPress={() => toggleCalendar()}>
          <MaterialIcons name='arrow-back' size={25} color='white' />
        </Pressable>
        <FontDisplay style={styles.title}>Calendar</FontDisplay>
      </View>
      {/* <Pressable onPress={() => getAllMonths()}>
        <FontDisplay style={styles.all}>ALL MONTHS</FontDisplay>
      </Pressable> */}
      <View style={styles.keyContainer}>
        <View style={[styles.key, styles.keyCurrent]}>
          <FontDisplay style={styles.keyText}>TODAY'S DATE</FontDisplay>
        </View>
        <View style={[styles.key, styles.keySelected]}>
          <FontDisplay style={styles.keyText}>SELECTED DATE</FontDisplay>
        </View>
      </View>

      {/* FLATLIST STUFF */}
      {/* <FlatList
        numColumns={3}
        data={DATA}
        renderItem={renderItem}
        keyExtractor={(item) => item.id}
        extraData={selectedId}
      /> */}

      <ScrollView>
        {months.map((month, index) => (
          <MonthItem
            key={index}
            id={month.id}
            name={month.name}
            days={month.days}
            onPress={() => handleMonth(month)}
          />
        ))}
        <View>
          {days.map((day, index) => (
            <DayItem
              key={index}
              id={day.id}
              name={day.name}
              onPress={() => handleDay(day)}
            />
          ))}
        </View>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    padding: 15
  },
  item: {
    padding: 5,
    marginVertical: 5,
    marginHorizontal: 5,
    width: '30%'
  },
  itemText: {
    textAlign: 'center'
  },
  controls: {
    alignItems: 'center',
    flexDirection: 'row',
    marginBottom: 30,
    marginTop: 15
  },
  title: {
    fontSize: 30,
    flex: 1,
    letterSpacing: 1,
    textAlign: 'center',
    paddingRight: 25
  },
  keyContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    marginBottom: 30
  },
  backIcon: {
    width: 25
  },
  key: {
    width: '30%',
    paddingBottom: 5,
    borderStyle: 'solid',
    borderBottomWidth: 15
  },
  keyText: {
    textAlign: 'center'
  },
  keyCurrent: {
    borderBottomColor: '#067b84'
  },
  keySelected: {
    borderBottomColor: '#2c2cb9'
  }
});

export default Calendar;
