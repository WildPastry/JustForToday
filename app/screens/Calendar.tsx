import { Text, View } from '../../components/Themed';

const Calendar: React.FC = (): JSX.Element => {
  return (
    <View>
      <Text>Calendar</Text>
    </View>
  );
};

export default Calendar;
