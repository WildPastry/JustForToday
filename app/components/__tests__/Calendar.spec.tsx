import Calendar from '../layout/Calendar';
import { Provider } from 'react-redux';
import { makeStore } from '../../redux/store';
import renderer from 'react-test-renderer';

describe('<Calendar />', () => {
  it('renders correctly', () => {
    const store = makeStore();
    const tree = renderer
      .create(
        <Provider store={store}>
          <Calendar
            handleCalendarState={() => {
              true;
            }}
            handleCalendarChange={() => {
              true;
            }}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
