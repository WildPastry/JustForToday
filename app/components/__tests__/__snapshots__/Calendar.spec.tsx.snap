// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar /> renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 20,
          },
          {
            "height": 35,
          },
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "justifyContent": "center",
          "width": "25%",
        }
      }
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#0074c1",
            },
            {
              "fontSize": 27,
              "lineHeight": 39,
            },
          ]
        }
      >
        CURRENT DAY
      </Text>
    </View>
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          {
            "justifyContent": "center",
            "width": "25%",
          },
        ]
      }
    />
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "gap": 20,
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        {
          "marginBottom": 20,
          "marginTop": 30,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 20,
            "justifyContent": "center",
          },
        ]
      }
    />
  </View>
</View>
`;
